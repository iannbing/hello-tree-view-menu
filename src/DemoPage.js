import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { get } from 'lodash';

import TreeViewMenu from './components/TreeViewMenu';
import data from './data/spaces.json';

class DemoPage extends Component {
  static getDerivedStateFromProps(props) {
    const path = get(props, 'match.url');
    return { path };
  }

  state = { path: '' };

  navigate = path => {
    const { history } = this.props;
    history.push(path);
  };

  render() {
    const { path } = this.state;
    return <TreeViewMenu data={data} path={path} navigate={this.navigate} />;
  }
}

export default withRouter(DemoPage);
