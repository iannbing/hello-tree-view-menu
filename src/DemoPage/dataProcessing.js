/* eslint-disable import/prefer-default-export */

const getRootPage = dataEntry => dataEntry.content.pages.find(page => page.m);

const getCategories = ({ dataEntry, spaceName }) => {
  const {
    content: { categories, pages }
  } = dataEntry;
  return categories.map(category => {
    const categoryPage = pages.find(page => page.t === category);
    const nodeName = categoryPage.u.replace(spaceName, '').replace(/\//g, '');
    return {
      node: nodeName,
      title: category,
      url: categoryPage.u
    };
  });
};

export const dehydrate = data => {
  const dehydrated = data.reduce((accu, curr) => {
    const spaceName = curr.filename.replace('.json', '');
    const { categories, pages } = curr.content;
    const rootPage = getRootPage(curr);
    // const categories = getCategories({ dataEntry: curr, spaceName });
    return [
      ...accu,
      {
        node: spaceName,
        title: rootPage.t,
        url: rootPage.u
      },
      ...categories
    ];
  }, []);

  return dehydrated;
};

export const transpose = ({ data, navigate }) => {};
