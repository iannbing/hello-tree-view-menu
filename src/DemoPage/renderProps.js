import React from 'react';
import styled from 'react-emotion';
import {
  ListGroupItem as DefaultListGroupItem,
  ListGroup,
  InputGroup,
  InputGroupAddon,
  Input,
} from 'reactstrap';

const DEFAULT_PADDING = 1.25;
const LEVEL_SPACE = 1.25;
const ICON_SIZE = 1;
const ToggleIcon = ({ on }) => <div>{on ? '-' : '+'}</div>;

const ToggleIconContainer = styled('div')(({ level = 0 }) => ({
  position: 'absolute',
  left: `${ICON_SIZE + level * LEVEL_SPACE}rem`,
  width: `${ICON_SIZE}rem`,
  height: `${ICON_SIZE}rem`,
}));

const ListGroupItemWithPadding = styled(DefaultListGroupItem)(
  ({ level = 0 }) => ({
    paddingLeft: `${DEFAULT_PADDING + ICON_SIZE + level * LEVEL_SPACE}rem`,
    cursor: 'pointer',
  }),
);

export const renderItem = ({ hasSubItems, isOpen, level, label, ...props }) => (
  <ListGroupItemWithPadding level={level} {...props}>
    {hasSubItems && (
      <ToggleIconContainer level={level}>
        <ToggleIcon on={isOpen} />
      </ToggleIconContainer>
    )}
    {label || 'Unkown'}
  </ListGroupItemWithPadding>
);

export const renderList = ({ search, items }) => {
  const onSearch = e => {
    const { value } = e.target;
    search(value);
  };
  return (
    <>
      <InputGroup>
        <InputGroupAddon addonType="prepend">Search</InputGroupAddon>
        <Input onChange={onSearch} />
      </InputGroup>
      <ListGroup>{items}</ListGroup>
    </>
  );
};
