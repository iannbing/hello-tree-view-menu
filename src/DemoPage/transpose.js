import { merge } from 'lodash';

const createObjFromKeys = ({ obj = {}, keys, value }) => {
  if (keys.length === 0) return {};
  if (keys.length === 1) {
    obj[keys[0]] = value;
  } else {
    const key = keys.shift();
    obj[key] = obj[key] || {};
    obj[key].nodes = {
      ...obj[key].nodes,
      ...createObjFromKeys({
        obj: !obj[key].nodes ? {} : obj[key].nodes,
        keys,
        value,
      }),
    };
  }
  return obj;
};

const cleanPath = path =>
  path
    .split('/')
    .filter(x => x)
    .join('/');

const transposeSpace = ({ space, spaceIndex }) => {
  const { content } = space;
  return content.pages.reduce((allPages, currentPage, pageIndex) => {
    const label = currentPage.t;
    const url = currentPage.u;
    const isSpace = !!currentPage.m;
    const parentPage = currentPage.p;
    const path = url ? url.split('/').filter(x => x) : '';
    const key = url ? cleanPath(url) : '';

    // if it has a parent page, insert it to have complete nodes
    if (parentPage) path.splice(path.length - 1, 0, parentPage);

    const newObj = createObjFromKeys({
      keys: path,
      value: {
        label,
        key,
        index: isSpace ? spaceIndex : pageIndex,
      },
    });
    return merge(newObj, allPages);
  }, {});
};

const transpose = ({ data, index }) => {
  if (Array.isArray(data))
    return data.reduce((allSpaces, currentSpace, spaceIndex) => {
      const currentSpacePages = transposeSpace({
        space: currentSpace,
        spaceIndex,
      });
      return {
        ...allSpaces,
        ...currentSpacePages,
      };
    }, {});
  return transposeSpace({
    space: data,
    spaceIndex: index,
  });
};

export default transpose;
