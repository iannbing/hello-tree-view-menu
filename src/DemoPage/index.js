import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import { get } from 'lodash';

import TreeViewMenu from '../components/TreeViewMenu';
import data from '../data/spaces.json';
import transpose from './transpose';

const cleanPath = path =>
  path
    .split('/')
    .filter(x => x)
    .join('/');

const Icon = ({ on }) => (
  <div style={{ fontWeight: 'bold', position: 'absolute', left: -8 }}>
    {on ? '[-]' : '[+]'}
  </div>
);

class DemoPage extends Component {
  static getDerivedStateFromProps(props) {
    const path = get(props, 'location.pathname');

    return { path };
  }

  state = { path: '', treeData: null };

  componentDidMount() {
    this.processData();
  }

  processData = () => {
    const treeData = transpose({ data, navigate: this.navigate });
    this.setState({ treeData });
  };

  navigate = path => {
    const { history } = this.props;
    history.push(path);
  };

  render() {
    const { path, treeData } = this.state;
    const activeKey = cleanPath(path);
    return (
      <>
        {treeData && (
          <TreeViewMenu
            data={treeData}
            activeKey={activeKey}
            search
            toggleIcon={Icon} // remove this prop to see the default icon
          />
        )}
      </>
    );
  }
}

export default withRouter(DemoPage);
