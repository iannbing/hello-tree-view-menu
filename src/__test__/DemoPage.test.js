import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';

import DemoPage from '../DemoPage';

const mockData = [{ foo: 'bar' }];
const mockPath = '/foo';

jest.mock('../components/TreeViewMenu', () => 'mock-tree-view-menu');
jest.mock('../data/spaces.json', () => [{ foo: 'bar' }]);

describe('DemoPage', () => {
  it('should render correctly', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ pathname: mockPath, key: 'path' }]}>
        <Route path={mockPath} render={props => <DemoPage {...props} />} />
      </MemoryRouter>
    );

    expect(wrapper.find(DemoPage)).toMatchSnapshot();
  });

  it('should provide correct props for TreeViewMenu', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[{ pathname: mockPath, key: 'path' }]}>
        <Route path={mockPath} render={props => <DemoPage {...props} />} />
      </MemoryRouter>
    );

    expect(wrapper.find('mock-tree-view-menu').prop('data')).toEqual(mockData);
    expect(wrapper.find('mock-tree-view-menu').prop('path')).toEqual(mockPath);
  });
});
