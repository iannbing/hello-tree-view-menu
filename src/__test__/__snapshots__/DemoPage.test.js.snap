// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DemoPage should render correctly 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/foo",
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "1isj1v",
            "pathname": "/foo",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "1isj1v",
          "pathname": "/foo",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Route
      path="/foo"
      render={[Function]}
    >
      <withRouter(DemoPage)
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "1isj1v",
                "pathname": "/foo",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "1isj1v",
              "pathname": "/foo",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "1isj1v",
            "pathname": "/foo",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/foo",
            "url": "/foo",
          }
        }
      >
        <Route>
          <DemoPage
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "canGo": [Function],
                "createHref": [Function],
                "entries": Array [
                  Object {
                    "hash": "",
                    "key": "1isj1v",
                    "pathname": "/foo",
                    "search": "",
                    "state": undefined,
                  },
                ],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "index": 0,
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "key": "1isj1v",
                  "pathname": "/foo",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "key": "1isj1v",
                "pathname": "/foo",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/foo",
                "url": "/foo",
              }
            }
          >
            <mock-tree-view-menu
              data={
                Array [
                  Object {
                    "foo": "bar",
                  },
                ]
              }
              path="/foo"
            />
          </DemoPage>
        </Route>
      </withRouter(DemoPage)>
    </Route>
  </Router>
</MemoryRouter>
`;
